### PR TITLE
feat: add support for create tables through metadata

### PR DESCRIPTION
**Pull request by olavloite**
Adds support for the `createTable` modification type.

### CODE REVIEW

File: src/main/java/liquibase/ext/spanner/SpannerCreateTableGenerator.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    StringBuilder buffer = new StringBuilder(", PRIMARY KEY (");
+    buffer.append(
+        database.escapeColumnNameList(
+            StringUtils.join(statement.getPrimaryKeyConstraint().getColumns(), ", ")));
+    buffer.append(")");
+
+    String pk = buffer.toString();
+    String sql = res[0].toSql();
+    sql = sql.replace(pk, "");
+    // Append PRIMARY KEY (without the leading ,)
```
Comment (mescanne) [inline src/main/java/liquibase/ext/spanner/SpannerCreateTableGenerator.java:43]: You're re-generating the PRIMARY KEY part, stripping it, and then adding it back.
Do we validate if there's any primary key? Spanner can't handle this, but Liquibase is happy creating tables without it.

File: src/test/java/liquibase/ext/spanner/AbstractMockServerTest.java [hunk 2]
```diff
@@ -40,7 +50,29 @@
...
+      Statement.of("SELECT COUNT(*) FROM DATABASECHANGELOGLOCK");
+  static final Statement DELETE_FROM_DATABASECHANGELOGLOCK =
+      Statement.of("DELETE FROM DATABASECHANGELOGLOCK WHERE true");
+  static final Statement INSERT_DATABASECHANGELOGLOCK =
+      Statement.of("INSERT INTO DATABASECHANGELOGLOCK (ID, LOCKED) VALUES (1, FALSE)");
+  static final Statement SELECT_LOCKED =
+      Statement.of("SELECT LOCKED FROM DATABASECHANGELOGLOCK WHERE ID=1");
+  static final Statement ACQUIRE_LOCK =
+      Statement.of(
+          "UPDATE UPDATE DATABASECHANGELOGLOCK SET LOCKED = TRUE, LOCKEDBY = 'some-host', LOCKGRANTED = '2020-10-27 19:13:41.732' WHERE ID = 1 AND LOCKED = FALSE");
```
[the remark this answers is not in the document]
Comment (mescanne) [reply src/test/java/liquibase/ext/spanner/AbstractMockServerTest.java:63]: When I changed this to just a single UPDATE the tests all passed. Without it, they failed.

**Review by mescanne — CHANGES_REQUESTED** [review]
The mock testing is an interesting approach. Laborious, but I can see the benefit. It also allows you very precise testing. The only concern is the UPDATE UPDATE and that I needed to change that to pass the tests locally.

I'd like to add create table tests using the Spanner emulator (and real Spanner in CI) as well, but that's future.

File: src/test/java/liquibase/ext/spanner/AbstractMockServerTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (olavloite) [reply src/test/java/liquibase/ext/spanner/AbstractMockServerTest.java]: No, that was not intentional, and strangely enough the tests succeed when executed in Eclipse and Maven, but fail when executed using Gradle. I've fixed the mocked statement.

File: src/main/java/liquibase/ext/spanner/SpannerCreateTableGenerator.java (hunk 1 above)
Comment (olavloite) [reply src/main/java/liquibase/ext/spanner/SpannerCreateTableGenerator.java:58]: Good point, I've added a validation step for that.

Cloud Spanner requires a Primary Key to be defined, but the list of columns is allowed to be empty. That creates a table that may hold at most 1 row. It is however not possible to define such a table using the meta model of Liquibase.

**Review by mescanne — APPROVED** [review]

**Comment by mescanne** [comment]
Great. Let's get it merged and move on to all of the others!

Mark

**Pull request merged**